### PR TITLE
Add support for query operations

### PR DESCRIPTION
**Pull request by rjew3ll**
Add Query operation and demo program.
Add ssl_connect functions.

This is just the query operation.  Still need to look at the tests.

### CODE REVIEW

File: kmip.c [hunk 1]
```diff
@@ -8377,6 +9054,76 @@ kmip_compare_response_message(const ResponseMessage *a, const ResponseMessage *b
...
+
+    return(KMIP_TRUE);
+}
+
+int
+kmip_compare_query_request_payload(const QueryRequestPayload *a, const QueryRequestPayload *b)
+{
+    (void) a;
+    (void) b;
+    return(KMIP_NOT_IMPLEMENTED);
```
Comment (PeterHamilton) [inline kmip.c:9115]: This needs to be implemented.

File: kmip.c [hunk 2]
```diff
@@ -8377,6 +9054,76 @@ kmip_compare_response_message(const ResponseMessage *a, const ResponseMessage *b
...
+}
+
+int
+kmip_compare_query_response_payload(const QueryResponsePayload *a, const QueryResponsePayload *b)
+{
+    (void) a;
+    (void) b;
+    return(KMIP_NOT_IMPLEMENTED);
```
Comment (PeterHamilton) [inline kmip.c:9123]: This needs to be implemented.

File: kmip.c [hunk 3]
```diff
@@ -10623,6 +11378,62 @@ kmip_encode_response_message(KMIP *ctx, const ResponseMessage *value)
...
+    return(KMIP_OK);
+}
+
+
+int
+kmip_encode_query_response_payload(KMIP *ctx, const QueryResponsePayload *value)
+{
+    (void) ctx;
+    (void) value;
+    return(KMIP_NOT_IMPLEMENTED);
```
Comment (PeterHamilton) [inline kmip.c:11434]: This needs to be implemented.

File: kmip.c [hunk 4]
```diff
@@ -13218,3 +14041,248 @@ kmip_decode_response_message(KMIP *ctx, ResponseMessage *value)
...
+
+    return(KMIP_OK);
+}
+
+int
+kmip_decode_query_request_payload(KMIP *ctx, QueryRequestPayload *value)
+{
+    (void) ctx;
+    (void) value;
+    return(KMIP_NOT_IMPLEMENTED);
```
Comment (PeterHamilton) [inline kmip.c:14090]: This needs to be implemented.

File: kmip.h [hunk 5]
```diff
@@ -669,6 +751,15 @@ enum tag
     KMIP_TAG_SERVER_CORRELATION_VALUE         = 0x420106,
     /* KMIP 2.0 */
     KMIP_TAG_ATTRIBUTES                       = 0x420125,
+    KMIP_TAG_SERVER_NAME                      = 0x42012D,  /* KMIP 2.0 */
```
Comment (PeterHamilton) [inline kmip.h:754]: Remove the trailing comment.

File: kmip.h [hunk 6]
```diff
@@ -1092,6 +1183,84 @@ typedef struct response_message
...
+ // Capability Information        No, MAY be repeated  // Specifies the capabilities that are supported by the server.
+ // Client Registration Method    No, MAY be repeated  // Specifies a Client Registration Method that is supported by the server.
+ // Defaults Information          No                   // Specifies the defaults that the server will use if the client omits them.
+ // Protection Storage Masks      Yes                  // Specifies the list of Protection Storage Mask values supported by the server. A server MAY elect to provide an empty list in the Response if it is unable or unwilling to provide this information.
+} QueryResponsePayload;
+
+
+#define MAX_QUERY_LEN    128
+#define MAX_QUERY_OPS   0x40
+#define MAX_QUERY_OBJS  0x20
```
Comment (PeterHamilton) [inline kmip.h:1244]: Move these `#define` statements over with the others.

File: kmip.h [hunk 7]
```diff
@@ -1092,6 +1183,84 @@ typedef struct response_message
...
+    bool32           server_information_valid;
+    char             server_name[MAX_QUERY_LEN];
+    char             server_serial_number[MAX_QUERY_LEN];
+    char             server_version[MAX_QUERY_LEN];
+    char             server_load[MAX_QUERY_LEN];
+    char             product_name[MAX_QUERY_LEN];
+    char             build_level[MAX_QUERY_LEN];
+    char             build_date[MAX_QUERY_LEN];
+    char             cluster_info[MAX_QUERY_LEN];
+} QueryResponse;
```
Comment (PeterHamilton) [inline kmip.h:1262]: The fields in this struct don' correspond to the fields defined in the specification. Defining maximum array lengths also doesn't align with the repeating nature of some of these fields. Most of them should be linked lists that grow as needed to fit the encoding or user usage.

File: ssl_connect.c [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    return bio;
+}
+
+void ssl_disconnect(BIO* bio)
+{
+    BIO_ssl_shutdown(bio);
+    BIO_free_all(bio);
+}
+
+
```
Comment (PeterHamilton) [inline ssl_connect.c:180]: For now separate out the Query changes from the SSL modifications. They should be evaluated separately.

File: ssl_connect.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+extern char ssl_saved_error[];
+extern int  ssl_reused;
+extern int  ssl_error;
+extern int  ssl_error_type;  // 1=server, 2=client
+
+void ssl_initialize();
+const char* openssl_strerror();
+SSL_CTX* ssl_create_context(char *client_certificate, char *client_key, char *ca_certificate);
+BIO*     ssl_connect(SSL_CTX *ctx, const char* server_address, const char* server_port, SSL_SESSION** session);
+void ssl_disconnect(BIO* bio);
```
Comment (PeterHamilton) [inline ssl_connect.h:21]: For now separate out the Query changes from the SSL modifications. They should be evaluated separately.

**Review by PeterHamilton — CHANGES_REQUESTED** [review]
Overall I like most of the Query changes. I'll need to do a second pass to verify the correctness on most of the new functionality. Unit tests must be added for most of the new functions and existing unit tests should be modified for the existing functions that were modified.

I am less sure if we want to use the SSL modifications. The demo applications are just that, demonstrations of how to use the library. If a developer wants to augment or improve how SSL is handled, then they can incorporate that into their own applications.

File: Makefile [hunk 10]
```diff
@@ -72,15 +73,17 @@ uninstall_html_docs:
 docs: html_docs
 html_docs:
 	cd $(SRCDIR)/docs && make html && cd -
-demos: demo_create demo_get demo_destroy
+demos: $(DEMOS)
 demo_get: demo_get.o $(OFILES)
-	$(CC) $(LDFLAGS) -o demo_get $? $(LDLIBS)
+	$(CC) $(LDFLAGS) -o demo_get $^ $(LDLIBS)
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply Makefile:78]: '$?'     The names of all the prerequisites that are newer than the target...
'$^'    The names of all the prerequisites ...

The link should include all the OFILES and should not depend on their timestamps.

File: Makefile [hunk 11]
```diff
@@ -100,12 +103,15 @@ kmip_memset.lo: kmip_memset.c kmip_memset.h
 kmip_bio.o: kmip_bio.c kmip_bio.h
 kmip_bio.lo: kmip_bio.c kmip_bio.h
 
+ssl_connect.o: ssl_connect.c ssl_connect.h
+ssl_connect.lo: ssl_connect.c ssl_connect.h
+
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply Makefile:108]: removed ssl wrapper

File: ssl_connect.c (hunk 8 above)
Comment (rjew3ll) [reply ssl_connect.c:180]: removed ssl wrapper

File: ssl_connect.h (hunk 9 above)
Comment (rjew3ll) [reply ssl_connect.h:21]: removed ssl wrapper

File: demo_query.c [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+    {
+        kmip_destroy(ctx);
+        return(KMIP_MEMORY_ALLOC_FAILED);
+    }
+    printf("encoding = %p\n", encoding);
+    kmip_set_buffer(ctx, encoding, buffer_total_size);
+
+    /* Build the request message. */
+
+
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:133]: done

File: demo_query.c [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+    kmip_print_request_message(stdout, &rm);
+    printf("\n");
+
+    char *response = NULL;
+    int response_size = 0;
+
+    printf("bio query send request\n");
+
+    int result = kmip_bio_send_request_encoding(ctx, bio, (char *)encoding, ctx->index - ctx->buffer, &response, &response_size);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:221]: done

File: demo_query.c [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+
+    LinkedList *functions = ctx->calloc_func(ctx->state, 1, sizeof(LinkedList));
+    for(size_t i = 0; i < query_count; i++)
+    {
+        LinkedListItem *item = ctx->calloc_func(ctx->state, 1, sizeof(LinkedListItem));
+        item->data = &queries[i];
+        kmip_linked_list_enqueue(functions, item);
+    }
+
+    //printf("functions = %p\n", (void*) functions);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:153]: done

File: demo_query.c [hunk 15]
```diff
@@ -0,0 +1,473 @@
...
+    printf("Result: ");
+    kmip_print_result_status_enum(stdout, result);
+    printf(" (%d)\n\n", result);
+
+    if(result == KMIP_STATUS_SUCCESS)
+    {
+        kmip_copy_query_result(query_result, (QueryResponsePayload*) req.response_payload);
+    }
+
+    //printf("bio query free response resp_m =  %p, response = %p\n", (void*)&resp_m, response);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:315]: done

File: demo_query.c [hunk 16]
```diff
@@ -0,0 +1,473 @@
...
+
+    /* Clean up the response message, the response buffer, and the KMIP */
+    /* context.                                                         */
+    kmip_free_response_message(ctx, &resp_m);
+    kmip_free_buffer(ctx, response, response_size);
+    response = NULL;
+
+    //kmip_set_buffer(ctx, NULL, 0);
+   // kmip_destroy(ctx);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:324]: done

File: demo_query.c [hunk 17]
```diff
@@ -0,0 +1,473 @@
...
+    enum query_function queries[] =
+    {
+        KMIP_QUERY_OPERATIONS,
+        KMIP_QUERY_OBJECTS,
+        KMIP_QUERY_SERVER_INFORMATION,
+        KMIP_QUERY_APPLICATION_NAMESPACES,
+    };
+
+    int result = kmip_bio_query_with_context(&kmip_context, bio, queries, ARRAY_LENGTH(queries), query_result);
+    //int result = use_low_level_api(&kmip_context, bio, queries, ARRAY_LENGTH(queries), query_result);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:354]: done

File: demo_query.c [hunk 18]
```diff
@@ -0,0 +1,473 @@
...
+int
+use_mid_level_api(BIO* bio,
+                  QueryResponse* query_result)
+{
+    /* Set up the KMIP context and send the request message. */
+    KMIP kmip_context = {0};
+
+    //kmip_context.calloc_func = &demo_calloc;
+    //kmip_context.realloc_func = &demo_realloc;
+    //kmip_context.free_func = &demo_free;
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:341]: done

File: demo_query.c [hunk 19]
```diff
@@ -0,0 +1,473 @@
...
+main(int argc, char **argv)
+{
+    char *server_address = NULL;
+    char *server_port = NULL;
+    char *client_certificate = NULL;
+    char *client_key = NULL;
+    char *ca_certificate = NULL;
+    int help = 0;
+    
+    int error = parse_arguments(argc, argv, &server_address, &server_port, &client_certificate, &client_key, &ca_certificate, &help);
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply demo_query.c:402]: done

File: kmip.c [hunk 20]
```diff
@@ -1132,6 +1134,56 @@ kmip_check_enum_value(enum kmip_version version, enum tag t, int value)
...
+            return(KMIP_OK);
+            break;
+        /* KMIP 1.1 */
+        case KMIP_QUERY_EXTENSION_LIST:
+        case KMIP_QUERY_EXTENSION_MAP:
+            if(version >= KMIP_1_1)
+                return(KMIP_OK);
+            else
+                return(KMIP_INVALID_FOR_VERSION);
+            break;
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply kmip.c:1154]: done

File: kmip.c [hunk 21]
```diff
@@ -1132,6 +1134,56 @@ kmip_check_enum_value(enum kmip_version version, enum tag t, int value)
...
+        /* KMIP 1.2 */
+        case KMIP_QUERY_ATTESTATION_TYPES:
+            if(version >= KMIP_1_2)
+                return(KMIP_OK);
+            else
+                return(KMIP_INVALID_FOR_VERSION);
+            break;
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply kmip.c:1161]: done

File: kmip.c [hunk 22]
```diff
@@ -1132,6 +1134,56 @@ kmip_check_enum_value(enum kmip_version version, enum tag t, int value)
...
+        case KMIP_QUERY_RNGS:
+        case KMIP_QUERY_VALIDATIONS:
+        case KMIP_QUERY_PROFILES:
+        case KMIP_QUERY_CAPABILITIES:
+        case KMIP_QUERY_CLIENT_REGISTRATION_METHODS:
+            if(version >= KMIP_1_3)
+                return(KMIP_OK);
+            else
+                return(KMIP_INVALID_FOR_VERSION);
+            break;
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply kmip.c:1172]: done

File: kmip.c [hunk 23]
```diff
@@ -1132,6 +1134,56 @@ kmip_check_enum_value(enum kmip_version version, enum tag t, int value)
...
+        /* KMIP 2.0 */
+        case KMIP_QUERY_DEFAULTS_INFORMATION:
+        case KMIP_QUERY_STORAGE_PROTECTION_MASKS:
+            if(version >= KMIP_2_0)
+                return(KMIP_OK);
+            else
+                return(KMIP_INVALID_FOR_VERSION);
+            break;
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply kmip.c:1180]: done

File: kmip.c [hunk 24]
```diff
@@ -1953,15 +2005,219 @@ kmip_print_operation_enum(FILE *f, enum operation value)
         case KMIP_OP_CREATE:
         fprintf(f, "Create");
         break;
-        
+
+        case KMIP_OP_CREATE_KEY_PAIR:
+        fprintf(f, "Create_key_pair");
```
[the remark this answers is not in the document]
Comment (rjew3ll) [reply kmip.c:2010]: done

**Review by PeterHamilton — CHANGES_REQUESTED** [review]
Thanks for updating this pull request. While I would prefer the not implemented functions to be filled in, the surrounding additions are definitely worthwhile and I would like to see them added to the library.

That said, at a minimum the new encoding/decoding/compare functions **must** have unit tests associated with them. See [here](https://github.com/OpenKMIP/libkmip/blob/master/tests.c#L6233) for an example of how the Create request payload encoding is tested.

I will not be able to merge this in until unit tests are added for:
- `kmip_compare_query_functions`
- `kmip_encode_query_functions`
- `kmip_encode_query_request_payload`
- `kmip_decode_query_functions`
- `kmip_decode_operations`
- `kmip_decode_object_types`
- `kmip_decode_server_information`
- `kmip_decode_query_response_payload`

Please also address the remaining comments on the `kmip.h` modifications. Thanks!

**Comment by rjew3ll** [comment]
I should be able to work on this again soon.

**Comment by rjew3ll** [comment]
I'm still working on this.  This is an intermediate drop with 4 tests added.

**Comment by srgg** [comment]
With all the respect to the efforts, any chance to get this done?

**Comment by PeterHamilton** [comment]
@AndrzejJakowski Thanks again for posting an update to this patch. I'm swamped with other things right now but I am hoping to spend some time on this within the next week.

@srgg Yes, I'm hoping to get this reviewed and merged in soon. My apologies for the delay!

**Comment by rjew3ll** [comment]
I think I've added the requested tests.   
The only thing I still need to do up update the way the data is returned to the caller.   The structure I used was intended to be simple and not require any memory management, but I agree it cannot handle all the fields added since KMIP 1.3.  (I just wanted the vendor name and build info).  If the data structure needs to be dynamic, the caller is going to have to help configure/allocate/free the memory needed.

File: tests.c [hunk 25]
```diff
@@ -11901,6 +11901,53 @@ test_decode_response_header_kmip_2_0(TestTracker *tracker)
...
+    Functions expected = {0};
+    expected.function_list = &list_1;
+
+    Functions observed = {0};
+
+    int result = kmip_decode_query_functions(&ctx, &observed);
+    int comparison = kmip_compare_query_functions(&expected, &observed);
+    if(!comparison)
+    {
+        // comparision is expected to fail
```
Comment (PeterHamilton) [inline tests.c:11940]: Why is this expected to fail? If you want to test a failed comparison, you'll need to add an `else` case that reports a failed test if the expected query functions match the observed ones.

File: tests.c [hunk 26]
```diff
@@ -11901,6 +11901,53 @@ test_decode_response_header_kmip_2_0(TestTracker *tracker)
...
+test_compare_query_functions(TestTracker *tracker)
+{
+    TRACK_TEST(tracker);
+
+    uint8 encoding[32] = {
+        0x42, 0x00, 0x74, 0x05, 0x00, 0x00, 0x00, 0x04,
+        0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
+        0x42, 0x00, 0x74, 0x05, 0x00, 0x00, 0x00, 0x04,
+        0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00
+    };
```
Comment (PeterHamilton) [inline tests.c:11914]: You don't necessarily need to use an encoding/decoding scheme to test query function comparisons. You can just build the lists manually yourself and compare that way.

**Review by PeterHamilton — CHANGES_REQUESTED** [review]
Thank you for submitting the additional tests. This looks pretty well rounded and adds a lot of useful functionality.

Just address that one question on that final test case comparing query functions and this will be good to go.

File: tests.c (hunk 25 above)
Comment (rjew3ll) [reply tests.c]: I changed the test to do a normal comparison

File: tests.c (hunk 26 above)
Comment (rjew3ll) [reply tests.c]: I changed the test to use two lists.

**Review by PeterHamilton — APPROVED** [review]

**Pull request merged**